_posts/2018/2018-09-02-careful-with-present-when-using-activerecord-scopes.md

```diff
@@ -0,0 +1,61 @@
+---
+title: Careful with present? when using ActiveRecord scopes
+description: 'Mostly it’s fine, until your app goes boom.'
+layout: article
+category: ruby
+image:
+  base: '2018/careful-with-present-when-using-activerecord-scopes'
+  alt: 'Stadtbibliothek, Stuttgart, Deutschland'
+  source: 'https://unsplash.com/photos/PkbZahEG2Ng'
+  credit: 'Tobias Fischer'
+---
+
+When using Rails, we tend to use `#present?` in many places as it is added to _all_ Ruby objects by Rails. For example, you might use it in a view to check if an array has some elements before trying to loop through it and display some data.
+
+However using `#present?` on models with large datasets can create unforeseen problems.
+
+
+## Instead of…
+
+...using `#present?` on an Active Relation (a scope on an Active Record class)
+
+```ruby
+books = Book.recently_released
+if books.present?
+  books.paperbacks.each { |book| puts book.title }
+end
+```
+
+
+## Use…
+
+...`#any?` or `#exists?`
+
+```ruby
+books = Book.recently_released
+if books.any?
+  books.paperbacks.each { |book| puts book.title }
+end
+```
+
+
+## But why?
+
+When you call `#present?` it instantiates the objects into an array in memory before checking whether that array is empty or not. In most cases this will be fine but in extreme cases, say complex models or when using a large dataset, you might see very high memory usage and slow performing requests.
+
+If you use `#any?`, a different, simpler SQL command is run that doesn't build a large array of Active Record objects in memory.
+
+If the relation has already been used, meaning the objects are loaded into memory, there’s no penalty to using `#present?` as it uses the existing array rather than loading a new one.
+
+As always with changes to how your application uses the database, you’ll definitely want to monitor how the changes affect the real-life performance characteristics of _your_ application. You need to make sure the extra SQL request from the call to `#any?` isn't a worse performing solution than using `#present?` for your data.
+
+You _could_ even monkey-patch the default behaviour of `#present?` on Active Relations in Rails, to always use `#any?`, by using the [`attendance` gem](https://github.com/schneems/attendance). This is a bit of a blunt instrument though and you’ll want to watch the performance of your app very closely if you take the plunge.
```

Comment:
Is this for people who prefer writing `#present?` or think it's semantically better? Why use `attendance` instead of just using `#any?`?

Reply:
Saves hunting down every occurrence.

Reply:
I thought of that...but then I thought 'Find & Replace' — but maybe that's dangerous...as you might have some other occurrence of 'present?' that's not a method call? 🤔